File: codingninjas/CountSpecialNode.py
from queue import Queue
from sys import stdin

# Following is the class defined to represent the node of a generic tree
class TreeNode :
    def __init__(self, data) :
        self.data = data
        self.children = list()
        
    def numChildren(self) :
        return len(self.children)
    
    def getChild(self, index) :
        if index > len(self.children) :
            return None
        
        return self.children[index]

def dfs(node,s):
    if(not node): return 0
    res = 0
    if(node.data in s): return 0
    res+=1
    s.add(node.data)
    n = node.numChildren()
    for i in range(n):
        res +=dfs(node.getChild(i),s)
    s.remove(node.data)
    return res
    
    

def countSpecialNodes(root) :   
    return dfs(root,set())



'''----------------Utility Function-----------------'''
def takeInput() :
    data = list(map(int, stdin.readline().strip().split(" ")))
    index = 0
    
    pendingNodes = Queue()
    rootData = data[index]
    index += 1
    
    root = TreeNode(rootData)
    pendingNodes.put(root)
    
    while not pendingNodes.empty() :
        currentNode = pendingNodes.get()
        numChild = data[index]
        index += 1
        
        for i in range(numChild) :
            currenChild = data[index]
            index += 1
            
            childNode = TreeNode(currenChild)
            pendingNodes.put(childNode)
            currentNode.children.append(childNode)
            
    
    return root

#main
t = int(stdin.readline().strip())

while t > 0 :
	
	root = takeInput();
	print(countSpecialNodes(root))

	t -= 1



    

File: codingninjas/CountSpecialNode.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;
    class TreeNode<T> {
        T data;
        ArrayList<TreeNode<T>> children;

        public TreeNode(T data) {
            this.data = data;
            this.children = new ArrayList<>();
        }

        public int numChildren() {
            return this.children.size();
        }

        public TreeNode<T> getChild(int index) {
            if (index > this.children.size()) {
                return null;
            }

            return this.children.get(index);
        }
    }



public class Solution {

    static int dfs(TreeNode<Integer> node,Set<Integer> s){
        if(node == null) return 0;
        int res = 0;
        if(s.contains(node.data)) return 0;
        res+=1;
        s.add(node.data);
        int n = node.numChildren();
        for(int i=0;i<n;i++){
            res+=dfs(node.getChild(i),s);
        }
        s.remove(node.data);
        return res;
    }
    public static int countSpecialNodes(TreeNode<Integer> root) {
        return dfs(root, new HashSet<>());

    }
}